### PR TITLE
Updates codebase to Go 1.14, preps for Heroku deploy

### PR DESCRIPTION
**Pull request by krashanoff**
Big PR round two. This is a much more digestible PR -- though it does still touch a great number of files in the repo. Here are all the changes made:
* Formatting enforcement
    - This PR updates existing code to align with gofmt requirements.
* Switch our codebase to Go 1.14.
    - The changes that this entails are pretty small. Just makes our package imports absolute.
* Add a Heroku Procfile.
* Add a go.mod file for dependencies.
    - This is needed by Heroku.
* Uses environment variable **values** for credential retrieval and port setting, instead of a path to a JSON file.
    - This has been a long time coming, but we have it now. This PR makes the server get its Firebase credentials from the `$TLACFG` environment variable, and set its port from the `$PORT` variable.
    - If no TLA credentials are provided, the app experiences a runtime failure.
    - If there is no PORT variable set, it defaults to 8081 as always.
* Consistently uses uid and pid query variables instead of mixing and matching programId, userId, uid, pid.
    - I noticed we were mixing and matching programId, userId, uid, and pid query variables all over the place in the backend. This PR fixes that so that we use uid and pid consistently.
* Update GitHub Actions workflow to actually work (whoops!).

### CODE REVIEW

File: db/db.go [hunk 1]
```diff
@@ -18,32 +19,28 @@ type DB struct {
...
 	}
 
 	// set up the app through which our client will be
 	// acquired.
-	opt := option.WithCredentialsFile(path)
+	opt := option.WithCredentialsJSON([]byte(cfg))
 	app, err := firebase.NewApp(ctx, nil, opt)
+	if err != nil {
+		return nil, err
+	}
```
Comment (krashanoff) [inline db/db.go:37]: Updates OpenFromEnv to use the TLACFG envvar.

File: server.go [hunk 2]
```diff
@@ -59,9 +55,16 @@ func main() {
 
 	log.Printf("endpoints initialized.")
 
+	// check for PORT variable.
+	port := os.Getenv("PORT")
+	if port == "" {
+		log.Printf("no $PORT environment variable provided, defaulting to '%s'", DEFAULTPORT)
+		port = "8081"
+	}
+
```
Comment (krashanoff) [inline server.go:64]: Gets port to serve on from envvars; defaults to 8081 if not provided.

**Comment by tfukaza** [comment]
Looks good on my end :)

**Comment by krashanoff** [comment]
Thanks Tomo! Will be merging now then and moving on to prepping our first release.

**Comment by mizlan** [comment]
> This has been a long time coming, but we have it now. This PR makes the server get its Firebase credentials from the `$TLACFG` environment variable, and set its port from the `$PORT` variable.

@krashanoff Why is this a good thing? Doesn't it make it a bit more annoying to test when running `go test`? The README also specifies use of a credentials JSON file as opposed to an environment variable. Could this be because the environment variable is easier managed by dotenv-esque tooling?

**Comment by mizlan** [comment]
Should there be a note that to run tests, this line must be ran first?

```bash
export TLACFG="$(cat credentials.json)"
```

**Comment by krashanoff** [comment]
Hi Michael,

Your point about "dotenv-esque tooling" is close to the mark. If I remember correctly, these tests required a Firebase connection to run (notice `Db` is not an interface type). To make them work in CI, I think we had saved the test environment's credentials into the repo's secrets. Your note about `export TLACFG=...` is correct. When I was working on this, I just copied the contents of the credentials file into my `.env`, delimited by single quotes so bash would handle the JSON nicely.

I agree that use of a `TLACFG` variable holding JSON credentials makes running and testing cumbersome. If I had to write the changes described here again, I'd probably just shove everything into a single config file or pull straight from the `credentials.json` itself, using the environment as a fallback.

I'd also recommend lifting database functionality to an interface and ensure tests run with a mock. Then, if the credentials field in the application's config is missing in a development environment, the program can switch to a degraded mode of operation using the mock.

Happy to answer other questions, but my contributions are over three years old now and have not crossed my mind in some time. Out of curiosity, what has got you looking into these old changes?

Leo

**Comment by mizlan** [comment]
Gotcha! There's good news though: other maintainers (particularly Timothy to my knowledge) have been working on this exact issue. There is already an interface (`TLADB`) which `db.DB` and `db.MockDB` both implement. Would it better to have 100% of tests use the mock, or would a mix of both be good? (I suppose that having Firebase tests can be used to test the correctness our mock implementation, by comparing the two behaviors.)

> Happy to answer other questions, but my contributions are over three years old now and have not crossed my mind in some time. Out of curiosity, what has got you looking into these old changes?

I'm the new maintainer, as Timothy, the previous maintainer, has graduated! This past quarter I've been familiarizing myself with the frontend codebase, but now that I have a pretty good mental model of everything going on in it, I thought I'd dig into the backend code too.

My plans for the near(ish) future are:
- fully migrate the functions that interact with the database from `db/` to `handler/` to allow for a mock to be used in place (since `handler/` functions works with `TLADB`s while `db/` functions work with `db.DB`s).
  - add some abstraction for supporting transactions
- finish a viable implementation of classes

**Comment by krashanoff** [comment]
Michael,

In my opinion, a mixture of two would be best, but frontloading more tests to use the mock seems like a better investment. I think Firebase is pretty straightforward, so any logic errors will be easy to catch ad-hoc if you use it while developing new features locally. Of course, writing proper E2E tests would be perfect, but it's all in the balance.

Glad that you're taking the initiative as the new maintainer. Your short-term plans sound well-scoped. Feel free to reach out if you have questions about the project's early history, and I can do my best to answer them.

Leo

**Pull request merged**